File: Javascript/1.Agosto/20.ArrayDuplicates-30Agosto.js
//todo Dato un array di interi, restituisci un array di interi che compaiono più di una volta nell'array iniziale, ordinati in ordine crescente. Se nessun valore compare più di una volta, restituisci un array vuoto.

//!Includi solo un'istanza di ciascun valore nell'array restituito.

//? Tempo di esecuzione:
//* 8 Minuti e 36 Secondi 
//?Codice scritto senza utilizzo di ChatGpt

function findDuplicates(arr) {

  let arraySort = arr.sort((a,b) => a - b);
  let arrayFinal = [];
  let isEmpty = true;

  for (let i = 0; i < arraySort.length; i++) {
    if(arraySort[i] === arraySort[i+1] && !arrayFinal.includes(arraySort[i])) {
     arrayFinal.push(arraySort[i])
     isEmpty = false
    }
  }

  if(isEmpty) return []
  else return arrayFinal;
}

console.log(findDuplicates([1, 2, 3, 4, 5]))

//?Codice scritto da ChatGpt

//*Non usata

//* Logica corretta

//* Eviti duplicati nel risultato

//* Ordinamento usato in modo sensato

//* Codice leggibile

//! sort modifica l’array originale (side effect)

//! isEmpty è inutile: puoi tornare direttamente l’array

//! includes rende l’algoritmo meno efficiente

//! L’ultimo confronto (i+1) va fuori array (JS lo gestisce, ma è sporco)

//? Versione Chat Gpt 

function findDuplicates(arr) {
  const sorted = [...arr].sort((a, b) => a - b);
  const result = [];

  for (let i = 0; i < sorted.length - 1; i++) {
    if (sorted[i] === sorted[i + 1]) {
      result.push(sorted[i]);
      i++; // salta i duplicati successivi
    }
  }

  return result;
}


//todo Considerazioni: 

//! N.B.

File: Python/1.Agosto/20.Array_duplicates-30Agosto.py
# //todo Consegna dell'esercizio
# Dato un array di interi, restituisci un array di interi che compaiono più di una volta nell'array iniziale,
# ordinati in ordine crescente. Se nessun valore compare più di una volta, restituisci un array vuoto.
# L'array restituito deve includere una sola occorrenza di ciascun valore.


# //* Codice scritto senza utilizzo di ChatGPT
def find_duplicates(arr):
    sort_list = sorted(arr)
    list_duplicates = []

    for indice in range(len(sort_list) - 1):
        if sort_list[indice] == sort_list[indice + 1]:
            if sort_list[indice] not in list_duplicates:
                list_duplicates.append(sort_list[indice])

    return list_duplicates


# //* Codice scritto da ChatGPT (correzione)
def find_duplicates_correct(arr):
    seen = set()
    duplicates = set()

    for num in arr:
        if num in seen:
            duplicates.add(num)
        else:
            seen.add(num)

    return sorted(list(duplicates))


# //* Pro e contro del mio codice

# //* Pro
# - Ottima idea usare sorted() per semplificare il confronto
# - Logica corretta nel confronto tra elementi consecutivi
# - Gestione corretta dell’unicità nel risultato

# //! Contro
# - Controllo "not in list" poco efficiente
# - Dipendenza dal sorting anche quando non strettamente necessario
# - Nomi variabili migliorabili (sort_list → sorted_arr)

# //* Consigli
# - Usa set quando devi solo verificare presenza o duplicati
# - Il tuo approccio va benissimo per esercizi base (ottimo per imparare)
# - Se vuoi migliorare performance, evita sorting quando non richiesto